test/reduce.test.js

```diff
@@ -177,6 +177,304 @@ const assertComparisonNotReduced = (t, comparison, value = 'test') => {
+            target: 'user.customAttributes.nameSuffix',
+          },
+          'resource.positions': {
+            comparison: 'subset',
+            target: 'user.customAttributes.positions',
+          },
+          'user.customAttributes.favoriteSauce': {
+            comparison: 'in',
+            value: ['ketchup', 'mayo'],
+          },
+        },
+      ],
+    },
+  };
+
+  t.deepEqual(reduce(initialPolicy, {}), expectedPolicy);
+});
+
+test('reverses conditions with custom user attributes as keys and evaluates policy correctly', (t) => {
+  const user = {
+    customAttributes: {
+      actions: ['HIKE', 'SNOWBOARD'],
+      group: 'pro-hikers elite',
+      rank: '1-2-3',
+      id: 'user-id',
+      isActive: true,
+      name: 'John Doe',
+      nameSuffix: 'The Third',
+      orgName: 'LifeOmic',
+      patient: 'patient-zero',
+      patients: ['patient-zero', 'patient-one', 'patient-two', 'patient-three'],
+      permissions: ['READ', 'WRITE'],
+      positions: ['BOSS'],
+      secret: 'confidential',
+    },
+  };
+  const matchingResource = Object.assign({}, user.customAttributes, {
+    allowedSuffix: 'The Third',
+    forbiddenAction: 'leave',
+    forbiddenNames: ['conor'],
+    orgName: 'Acme',
+    parentGroup: 'pro-hikers',
+    rankOrder: '1-2-3-4-5-6-7',
+  });
+  const privateResource = {
+    actions: ['HIKE'],
+    allowedSuffix: 'The Third',
+    forbiddenAction: 'leave',
+    forbiddenNames: ['conor'],
+    group: 'pro skiers',
+    id: 'resource-id',
+    isActive: false,
+    name: 'Mac Miller',
+    nameSuffix: 'Knight The Third',
+    orgName: 'Acme',
+    parentGroup: 'pro-hikers',
+    patient: 'patient-zero',
+    patients: ['patient-x'],
+    permissions: ['READ'],
+    positions: ['BOSS', 'VP'],
+    rankOrder: '1-2',
+    secret: 'super-secret-stuff',
+  };
+
+  const initialPolicy = {
+    rules: {
+      // Not expected to be in final policy since the custom conditions
+      // are not supposed to match.
+      writeData: [getUserCustomConditions('privateResource')],
+      // Expected to be true in the final policy since the custom conditions
+      // are supposed to match.
+      readData: [getUserCustomConditions('matchingResource')],
+      // Expected to not be evaluated and just have literal policy values
+      deleteData: [
+        getUserCustomConditions('privateResource'),
+        {
+          'user.customAttributes.favoriteSauce': {
+            comparison: 'in',
+            value: ['ketchup', 'mayo'],
+          },
+        },
+      ],
+    },
+  };
+  const expectedPolicy = {
+    rules: {
+      readData: true,
+      deleteData: [
+        {
+          'user.customAttributes.favoriteSauce': {
+            comparison: 'in',
+            value: ['ketchup', 'mayo'],
+          },
+        },
+      ],
+    },
+  };
+
+  t.deepEqual(
+    reduce(initialPolicy, {
+      user,
+      matchingResource,
+      privateResource,
+    }),
+    expectedPolicy
+  );
```

Comment:
I wonder, should we add an additional "end-to-end" assertion that confirms the operations are enforced the same way between the original + reduced policies?

Something like (pseudocode):

```typescript
const original = {
  rules: {
    includesRule: [
      {
         'testValue': {
           comparison: 'includes',
           target: 'someTarget'
         }
      }
    ]
  }
}

const reduced = reduce(original, { testValue: 'blah' })

expect(
  enforce(original, 'includesRule', { testValue: 'blah', someTarget: 'blah' })
).toStrictEqual(
  enforce(reduced, 'includesRule', { someTarget: 'blah' }))
```

Maybe we could even programmatically make that assertion for each comparison type?

Reply:
@swain let me know if you think that what I added is enough. I wasn't sure if you were recommending to basically have one rule for each specific comparison type and then assert on each one individually.